### DISCUSSION TITLE
How to use display (wip)

### DISCUSSION THREAD

**Discussion by smougenot** [arendst/Tasmota#11484] [Show and tell]
# How to use display

It is prety convenient to use a device with a displays. And Tasmota can manage th display.
This documentation is base on Tasmota 9.3.1
The device used is TTGO-displays
* ESP32 base
* With a display managed by `ST7789`

This will show
* howto use predefined display firmware for ESP32 using Platformio
* howto use custom firmware with Platformio //TODO
* howto configure using console with Platformio
* howto display time
* howto display sensor data //TODO

Requirements:
* a device + usb cable
* Platformio
* Git to get original sources code and manage we change (or download the source code zip from github)
// TODO link

## Devide 

TTGO-displays
* ESP32 base
* Integrates a display managed by `ST7789`
* relatively cheap

Pinout is here 
![pinmap](https://user-images.githubusercontent.com/1217669/112716881-40c35780-8ee9-11eb-9c17-c3454748e588.jpg)

//TODO link

## Platformio

IMOO: the easiest way to get going with Tasmota and ESP32 is using Platformio. 
* it is integrated with Visual Studio Code coding toolbox
* it is free and open source
* with very few command line (or clicks) we will be done.
* it allows us to customize the firmware (we'ii need this to display sensor data)

### Install the tool
Let see the official Platformio documentation the install instructions
//TODO: link

### Usefull commands

Platformio has a convenient `pio` command line tool.

#### Build firmware and upload to the device

```
pio run -t upload
```

![tasmota_display_firmware_32_Peek 2021-03-27 10-07](https://user-images.githubusercontent.com/1217669/112716897-65b7ca80-8ee9-11eb-9196-0362ecaa80a2.gif)

#### Acces to the console

In the console we will 
* get information displayed by Tasmota
* set configuration

```
pio device monitor --baud 115200
```
![tasmota_console_boot_Peek 2021-03-27 10-08](https://user-images.githubusercontent.com/1217669/112716910-77996d80-8ee9-11eb-863c-4d36f306cd85.gif)

Be careful the console does not display the characters we are typing ... it is normal.

### Use gith
// TODO
### Download from Github
// TODO

## Fix Platformio config

For reason explain in a comment below Tasmota team forces port definition in the Platformio config.
Unfortunately this prevents us from using device auto-detection feature of Platformio.
So if you do use only one ESP device at a time (most of us do) you are safe to enabler auto-detection.

To fix this auto-detection feature :
* comment line starting with `upload_port` (by adding a `;` at the beginning)
* several files are to be changed

Existing code
```
upload_port                 = COM5
```

What we want
```
;upload_port                 = COM5
```

## Using Display firmware

Tasmota team does a great job maitaining a lot of firmwares.
One of them is the `Display` flavour of the firmware.

We will build the firmaware and upload it to the device.

### Code preparation

Set Platformio configuration to build the Display firmware
* ensure the device is connected
* uncomment the line containing `tasmota32-display` in file named `platformio_tasmota32.ini` (by removing the  `;` at line start)

Existing code
```
;                tasmota32-display
```

What we want
```
                tasmota32-display
```

### Build and upload Firmware

see the usfull commands about firmware upload

### Confirm the Tasmota firmaware boots

see the usfull commands about console

## Configuring the device using the console

Of course we can do a lot using the web interface of Tasmota running on ou device.
But for using displays I had to use the console.
So let do all the configuration that way.

:warning: do not worry about command case Tasmota manages. (e.g. mqtthost or mqttHost will work both)

### Access the console

see the usfull commands about console

### Configuring wifi

We use tow Tasmota commands to set wifi
command     | usage 
----        | ----
ssid1       | set ssid of our network
password1   | set associated password

// TODO links

Either of thos command causes Tasmota to reboot.
We will send those two at once.

**Send the command**
Use a command like this ... with the correct information about the Wifi to use 
(remember those devices are week, it would be nice to place them in a dedicated wifi ... not the wifi used by the phone and the computer)

```
backlog ssid1 vulnerable_wifi_because_of_usage; password1 verry_complex_wifi_password
```

### Set up the MQTT (optional)

command         | usage 
----            | ----
mqtthost        | 
mqttuser        |
mqttpassword    |

// TODO links

```
backlog mqtthost pi4; mqttuser esp; mqttpassword very_complex_mqtt_password_indeed
```

### Enable DST

If we need DST, it requires activation

**send command to activate DST**
```
Timezone 99
```

### Set up the ports

Display features of Tasmota depends on the actual detection of a display.
This means that all the command are disabled until a valid display configuration of ports is set (Tasmota responds like the command does not exists)

One quick way to do the config is to find/build the appropriate tempalte for your device.
A lot of templates are referenced in //TODO link

For the TTGO-Display device a working template id 
```
 {"NAME":"ESP32-ttgo-display","GPIO":[32,1,1,1,992,6592,1,1,1,1,1,1,6624,1,896,864,0,1,1,1024,0,1,1,1,0,0,0,0,1,1,1,33,1,0,0,1],"FLAG":0,"BASE":1}
 ```

 **send the command**
```
template {"NAME":"ESP32-ttgo-display","GPIO":[32,1,1,1,992,6592,1,1,1,1,1,1,6624,1,896,864,0,1,1,1024,0,1,1,1,0,0,0,0,1,1,1,33,1,0,0,1],"FLAG":0,"BASE":1}
```

### Set up the display for TTGO-Display device

command         | usage 
----            | ----
DisplayWidth    | 
DisplayHeight   |
DisplayDimmer   |

// TODO links

```
backlog DisplayWidth 240; DisplayHeight 135; DisplayDimmer 15
```

### Display test

As a test we will display current time 

```
DisplayText [zs2tS]
```
![IMG_20210327_102921](https://user-images.githubusercontent.com/1217669/112716859-238e8900-8ee9-11eb-8eb0-f55b148ec4dd.jpg)

## Display time

If we want to display time, we want it to change every minute to reflect the current time.
In Tasmota this is done using Rules.

We will have to create a Rule that:
* display time ass soon as it is known (after boor, network access, clock setting)
* update time every minutes

**Configure the rule**
```
rule1 on Time#Initialized do DisplayText [l2c5C2047s2t] endon on Time#Minute do DisplayText [l2c5C2047s2t] endon
```

**Activate the rule**
```
rule1 on
```

**Deactivate the rule**
```
rule1 off
```

**After reboot you see this in the console**
![2021-03-27_10-36-44](https://user-images.githubusercontent.com/1217669/112716799-d1e5fe80-8ee8-11eb-8f71-56127ea04cc7.png)

**The display looks like**
![IMG_20210327_104104](https://user-images.githubusercontent.com/1217669/112716853-1bcee480-8ee9-11eb-8fc2-35bd94377788.jpg)

**Comment by Jason2866** [comment]
**Thx for this nice write up!**

Upload Port is fixed since we do not want to flash random targets ;-) So as intended.
And it is **Platformio** not Plateformio

**Comment by 2mar** [comment]
I tried to flash a TTGO -T display with Tasmota according to your instructions, but unfortunately the display remains dead except for the backlight. That's why I'm asking whether the display should work with the current tasmota32-display.bin version 10.1.0? If not could you please email me to ram2@gmx.net a working compiled version of tasmota for the TTGO T display? 
thank you very much

**Reply by jbeere** [reply]
You will need to compile for the TTGO esp32 with the following in your `user_config_override.h` file:

```
#define USE_DISPLAY
#define USE_DISPLAY_ST7789
#define USE_SPI
```

The display modes 1-5 are not implemented for the ST7789 (yet?), however you can test with the DisplayText commands like in this example.

You can ensure that the display is working by toggling the power on/off, the backlight is visible on the edges of the display.

edit:
I also found this useful for testing:

https://github.com/gemu2015/Sonoff-Tasmota/issues/28#issuecomment-751881775

**Reply by 2mar** [reply]
Thanks for the help!! Now the TTGO display modul works normal!

**Comment by robix62** [comment]
hi ! is there a firmware with tasmosta that I can upload with web installer ? I have some difficulties to use my esp32 ttgo display.
